backend/src/prereq_loader.ts

```diff
@@ -0,0 +1,250 @@
+import {
+  Schedule,
+  INEUAndPrereq,
+  INEUOrPrereq,
+  ScheduleYear,
+  ScheduleTerm,
+  ScheduleCourse,
+} from "../../frontend/src/models/types";
+import DataLoader from "dataloader";
+import request from "request-promise";
+
+/**
+ * Courses should have at least a classId and a subject, in order to query them.
+ * May have a termId.
+ */
+interface SimpleCourse {
+  termId?: number;
+  subject: string;
+  classId: number;
+}
+
+/**
+ * The result of a prereq query for a class.
+ */
+interface NonEmptyQueryResult {
+  prereqs?: INEUOrPrereq | INEUAndPrereq;
+  coreqs?: INEUOrPrereq | INEUAndPrereq;
+}
+
+type PrereqQueryResult = undefined | NonEmptyQueryResult;
+
+// keep track of the total number of requests made.
+let totalNumRequests = 0;
+
+/**
+ * Asynchronously adds prereqs to a Schedule.
+ * Does not do mutation.
+ * @param schedules the schedule to add prereqs to
+ */
+export async function addPrereqsToSchedules(
+  schedules: Schedule[]
+): Promise<Schedule[]> {
+  // set the total to zero
+  totalNumRequests = 0;
+
+  // the loader to use for building a
+  const loader = new DataLoader<SimpleCourse, PrereqQueryResult>(
+    (keys: SimpleCourse[]) => queryCoursePrereqData(keys)
+  );
+
+  // return the results
+  let results = await Promise.all(
+    schedules.map((sched: Schedule) => prereqifySchedule(sched, loader))
+  );
+
+  // log the total number of requests.
+  // console.log("total: " + totalNumRequests);
+  totalNumRequests = 0;
+
+  return results;
+}
+
+/**
+ * Loads prereqs for a given schedule
+ * @param schedule the schedule
+ * @param loader the loader
+ */
+async function prereqifySchedule(
+  schedule: Schedule,
+  loader: DataLoader<SimpleCourse, PrereqQueryResult>
+): Promise<Schedule> {
+  // does not do mutation!
+  const newYearMap: { [key: number]: ScheduleYear } = {};
+
+  // convert each of the years
+  for (const year of schedule.years) {
+    newYearMap[year] = await prereqifyScheduleYear(
+      schedule.yearMap[year],
+      loader
+    );
+  }
+
+  // reconstructs a year
+  return {
+    years: schedule.years.slice(0),
```

Comment:
isn't this the same as `years: schedule.years` ?

Reply:
Yes, almost. I had to do this because `schedule.years` for some reason didn't always have the actual `Array` prototype. Doing this converts from the fake proto to the real one.

That being said, I changed it back and nothing broke, so idk!